File: bindings/prolog/safe_path.pl
%% SPDX-License-Identifier: PMPL-1.0-or-later
%% Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
%%
%% Proven SafePath - Extended FFI bindings to libproven path operations.
%% All path operations are performed in verified Idris 2 code via libproven.

:- module(safe_path, [
    has_traversal/1,
    sanitize_filename/2,
    safe_path_join/3
]).

:- use_module(proven_path, [
    has_traversal/1 as pp_has_traversal,
    sanitize_filename/2 as pp_sanitize_filename,
    safe_path_join/3 as pp_safe_path_join
]).

%! has_traversal(+Path) is semidet.
%
%  Succeeds if Path contains directory traversal sequences.
%  Detection performed by libproven.
has_traversal(Path) :-
    pp_has_traversal(Path).

%! sanitize_filename(+Input, -Output) is det.
%
%  Sanitize a filename by removing dangerous characters via libproven.
sanitize_filename(Input, Output) :-
    pp_sanitize_filename(Input, Output).

%! safe_path_join(+Base, +Filename, -Result) is det.
%
%  Safely join base path and filename via libproven.
%  Result is ok(Path) or error(traversal_detected).
safe_path_join(Base, Filename, Result) :-
    pp_safe_path_join(Base, Filename, Result).

File: bindings/java/src/main/java/io/github/hyperpolymath/proven/SafePath.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
package io.github.hyperpolymath.proven;

import com.sun.jna.Memory;

import java.nio.charset.StandardCharsets;
import java.util.Optional;

/**
 * Safe filesystem path operations via libproven FFI.
 *
 * Provides path traversal detection and filename sanitisation.
 * All logic delegates to the Idris 2 verified implementation through
 * the Zig C ABI. No path-handling logic is reimplemented in Java.
 */
public final class SafePath {

    private static final ProvenLibrary LIB = ProvenLibrary.INSTANCE;

    private SafePath() {}

    /**
     * Check if a path contains directory traversal sequences (../, etc.).
     *
     * @param path the path to check
     * @return true if traversal is detected, empty on null pointer error
     */
    public static Optional<Boolean> hasTraversal(String path) {
        Memory mem = NativeUtil.toNativeString(path);
        if (mem == null) {
            return Optional.empty();
        }
        byte[] bytes = path.getBytes(StandardCharsets.UTF_8);
        return NativeUtil.extractBool(
            LIB.proven_path_has_traversal(mem, bytes.length));
    }

    /**
     * Sanitize a filename by removing dangerous characters.
     *
     * @param filename the filename to sanitize
     * @return the sanitized filename, or empty on error
     */
    public static Optional<String> sanitizeFilename(String filename) {
        Memory mem = NativeUtil.toNativeString(filename);
        if (mem == null) {
            return Optional.empty();
        }
        byte[] bytes = filename.getBytes(StandardCharsets.UTF_8);
        return NativeUtil.extractString(
            LIB.proven_path_sanitize_filename(mem, bytes.length));
    }
}
